File: models/gateway.go
package models

// HelloData is the payload for opcode 10 (Hello), sent by the server immediately after a WebSocket connection is
// established. The client must begin sending heartbeats at the specified interval.
type HelloData struct {
	HeartbeatInterval int `json:"heartbeat_interval"`
}

// IdentifyData is the payload for opcode 2 (Identify), sent by the client to authenticate a new connection. The token
// is a valid JWT access token.
type IdentifyData struct {
	Token string `json:"token"`
}

// ResumeData is the payload for opcode 6 (Resume), sent by the client to restore a previously disconnected session.
// The server replays any events with sequence numbers greater than seq.
type ResumeData struct {
	Token     string `json:"token"`
	SessionID string `json:"session_id"`
	Seq       int64  `json:"seq"`
}

// ReadyData is the payload for the READY dispatch event (opcode 0, type "READY"), sent after a successful Identify.
// It contains the initial state the client needs to populate its local cache.
type ReadyData struct {
	SessionID string       `json:"session_id"`
	User      User         `json:"user"`
	Server    ServerConfig `json:"server"`
	Channels  []Channel    `json:"channels"`
	Roles     []Role       `json:"roles"`
	Members   []Member     `json:"members"`
}

// MessageDeleteData is the payload for the MESSAGE_DELETE dispatch event, identifying the deleted message and the
// channel it belonged to.
type MessageDeleteData struct {
	ID        string `json:"id"`
	ChannelID string `json:"channel_id"`
}

File: models/gateway.ts
import type { Channel } from "./channel";
import type { Member } from "./member";
import type { Role } from "./role";
import type { ServerConfig } from "./server";
import type { User } from "./user";

/** Payload for opcode 10 (Hello), sent by the server on connection. */
export interface HelloData {
  heartbeat_interval: number;
}

/** Payload for opcode 2 (Identify), sent by the client to authenticate. */
export interface IdentifyData {
  token: string;
}

/** Payload for opcode 6 (Resume), sent by the client to restore a session. */
export interface ResumeData {
  token: string;
  session_id: string;
  seq: number;
}

/** Payload for opcode 3 (PresenceUpdate), sent by the client to set its own online status. */
export interface PresenceUpdateRequest {
  status: string;
}

/** Payload for the PRESENCE_UPDATE dispatch event, broadcast when a user's presence changes. */
export interface PresenceUpdateData {
  user_id: string;
  status: string;
}

/** Snapshot of a single user's presence, included in the READY payload. */
export interface PresenceState {
  user_id: string;
  status: string;
}

/** Payload for the TYPING_START dispatch event. */
export interface TypingStartData {
  channel_id: string;
  user_id: string;
  timestamp: string;
}

/** Payload for the READY dispatch event, containing initial client state. */
export interface ReadyData {
  session_id: string;
  user: User;
  server: ServerConfig;
  channels: Channel[];
  roles: Role[];
  members: Member[];
  presences: PresenceState[];
}

/** Payload for the MESSAGE_DELETE dispatch event. */
export interface MessageDeleteData {
  id: string;
  channel_id: string;
}

/** Payload for the CHANNEL_DELETE dispatch event. */
export interface ChannelDeleteData {
  id: string;
}

/** Payload for the ROLE_DELETE dispatch event. */
export interface RoleDeleteData {
  id: string;
}

/** Payload for the MEMBER_REMOVE dispatch event. */
export interface MemberRemoveData {
  user_id: string;
}
